Add archive import/export

hauntii opened the pull request:
This PR adds the following:

- Custom File and Debug Menu
- SARC and SZS import/export via File
- Window name is now `Ita Bag` instead of `Document`
- Preliminary function pathways for saving tabs
- App no longer throttles in background for background saving

![image](https://user-images.githubusercontent.com/57009359/214632921-b88d2365-fe64-4aa9-8116-f94be1f532a0.png)
![image](https://user-images.githubusercontent.com/57009359/214632933-4915b63d-7d5b-4d18-b8de-10fae7f387ac.png)

hauntii committed:
Support mega badges

Remove redundant code

hauntii committed:
Add archive import/export

jonbarrow commented:
I will do a proper code review after work but I'll leave this as some notes:

1. Was the saving and exporting actually tested? The official Badge Arcade sarc contains many additional sarc archives, and every non-xml file is YAZ0 compressed. Im not confident just slapping every file into one sarc uncompressed will work. Also the sarc file which holds the prize data inside the root sarc has a specific naming convention that must be handled, it's the dates the badges are active for

2. "Mega badges" are intentionally not supported because we do not handle the additional optional parts of the badge, both decoding and encoding, so removing those checks does not "add support" for them. That is completely unintentional and I will only accept a PR for mega badges if proper decoding/encoding of those additional parts is included

hauntii commented:
On 1), while it does load in the app, you are correct, as it does not load into Badge Arcade. I am working on a fix for this.

On 2) I'm most likely going to revert that commit and focus on importing/exporting.

hauntii committed:
Revert "Support mega badges"

This reverts commit 1795e79283c0743f77885a3b0a570859dd094fce.

hauntii committed:
Add compression

hauntii commented:
I've tested compressing the files, however they seem to get broken by yaz0lib. 
I have a suspicion it's that, as the Buffers for the files are the same, however the compressed data is way off.
Compressing with [yaz0dec](https://cdn.discordapp.com/attachments/163993370635665409/359538263733698583/Yaz0dec.rar) gets a closer result to the original files.
I've committed the code I currently have for compression if you'd like to take a stab at it.

jonbarrow commented:
> Compressing with [yaz0enc](http://amnoid.de/gc/yaz0enc.zip) gets a closer result to the original files.

Trying to download this zip file threw a sedurity warning in Edge, I'm not very comfortable with this being implemented here

hauntii commented:
It's less about implementing that exact program, it was more to have a running example of a closer implementation. There is a cpp source file inside the zip, and chrome did not throw a warning if you do want to look at the source. I can also provide the CPP if you want to look at it.

jonbarrow commented:
I also will say that I plan on changing how the virtual file system is stuctured soon so this export feature will need to be rewritten anyway. I would suggest holding off on that. The import feature is fine though

hauntii committed:
Fix export and add progress

hauntii commented:
As it turns out, the problem was that it was being saved as big endian rather than little endian.
I've fixed this and added a progress bar. (Ooh, fancy!)

Here it is in action:
![nice-fancy-progress-bar](https://user-images.githubusercontent.com/57009359/215393619-48651d42-3b35-4825-88c1-a6eba5d71624.gif)

While the vFS is getting rewritten, this is still a good base for a rewrite.

hauntii committed:
Fix save operation overlaying by disabling menu item

hauntii committed:
Remove original menu disable code

DaniElectra commented on app/assets/js/archive.js:
```diff
@@ -61,8 +55,37 @@ function sarcToObject(sarc, root) {
+            await new Promise(r => setImmediate(r)); // Wait for the UI to catch up
+
+			newPath = newPath.concat(newPath.endsWith('.xml') ? '' : '.szs');
+			sarc.addRawFile(newPath.endsWith('.xml') ? value : compressYaz0(value, 0, 1), newPath);
```
Although this works with the PrizeCollection folder `/sharc/pc`, this doesn't fully apply with a full archive, as there are more files aside from XML that aren't compressed (for example: MSBT dialogs, jpeg images and some raw images that don't use any extensions).

imo I would check to only compress files with the PrizeCollection extensions (prb, cab, cib...) instead of compressing everything except for some files, since those are the only files that are compressed

jonbarrow replied:
Do we even know if the files need to be compressed? The game might check if it's compressed first and not assume

DaniElectra replied:
The game refuses to load the file if it isn't compressed, so yeah (tested rn)

jonbarrow replied:
I've noticed that it seems like what files are compressed depends on the parent folder (every file in folder X is compressed), so maybe we should have a list of folders which need to be compressed?

I'm skeptical on doing things like this anyway, I'm thinking the packing of the data should be less "automatic" like the reading of it is, the packing should probably build each internal SARC one by one and add the compressed files to it, so the end result better matches the original data

DaniElectra replied:
The folder locations of the Yaz0-compressed files are all of the subfolders of a PrizeCollection (not including the `pc` folder), so we could use that to do the compression.

Also, the location of some directories can be changed inside the `Schedule.xml` (like the PrizeCollection and the Miiverse posts), so we could use it to also locate the folders that have to be packed as a SARC archive and pack them individually (I haven't played with the schedule, but this is in theory how it works)

hauntii replied:
I wouldn't recommend changing the directories, `allbadge.sarc` does not contain a `Schedule.xml` like the one inside of `boss.sarc` does. I think filetype compression would work just fine for this. The data packing being automatic should cause no issues with accuracy, as doing tests on importing and exporting causes exact binary duplicates to export.

hauntii replied:
Well, I've gotten almost everything working here. Only one issue, the SARCs inside of `/talkpic` seem to get corrupted either on compression or decompression. This seems to occur whether using `sarclib`, or `sarc_tool.exe`. Do either of you have any insight on this?

Recompressing them and replacing them into `boss.sarc` causes the game to immediately freeze up.

DaniElectra replied:
> Well, I've gotten almost everything working here. Only one issue, the SARCs inside of `/talkpic` seem to get corrupted either on compression or decompression. This seems to occur whether using `sarclib`, or `sarc_tool.exe`. Do either of you have any insight on this?
> 
> Recompressing them and replacing them into `boss.sarc` causes the game to immediately freeze up.

Sorry for the *very late* response.

If you were getting the images exported from here, you had got them compressed with Yaz0 `TpOb_00ZZ_AutomaticHallPrFreePlay_320x156_euES.szs` (because they pass the check that they aren't xml files and get compressed), even when they aren't in reality (should be `TpOb_00ZZ_AutomaticHallPrFreePlay_320x156_euES`). Decompressing the files gives the correct images

If you were getting the images uncompressed, it could be something about padding the data section of the SARC file.
From https://github.com/Kinnay/Nintendo-File-Formats/wiki/SARC-File-Format :
> Certain files, such as textures, have specific alignment requirements. For this reason, there may be padding between the SFNT and data section, and between individual files within the data section.

DaniElectra commented on index.js:
```diff
@@ -34,4 +39,98 @@ app.on('window-all-closed', () => {
+		id: 'export-archive',
+		label: 'Export Archive',
+		accelerator: 'CmdOrCtrl+E',
+		click: () => { handleFileSave(window, 'export-archive', { filters: [{name: 'PrizeCollection Archive', extensions: ['sarc', 'szs']}]}) }
```
I'm not sure what is the purpose of exporting archives as szs? None of the SARC archives are Yaz0 compressed

jonbarrow replied:
They are not, it should not be exportable to SZS

DaniElectra reviewed (commented):
Only some slight issues, but mostly lgtm

hauntii committed:
Disable szs as file export/import type

DaniElectra commented:
Considering we are going to change the vfs in the future, I think we can merge this PR and rewrite the export feature later. What do you think? @jonbarrow 

> I also will say that I plan on changing how the virtual file system is stuctured soon so this export feature will need to be rewritten anyway. I would suggest holding off on that. The import feature is fine though

jonbarrow commented:
Sounds good to me

hauntii committed:
Signing commit

jonbarrow commented:
I am going to force approve this, since you're part of the core team, however please be aware that going forward you should assume all repositories require signed commits
